Fix for bookmarks favicon caching

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/414709148257752/1202223699061973/f
CC: @miasma13 

**Description**: When visiting sites ensure that only favicons for bookmarked sites are saved to the bookmarks cache

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:
1. Visit a website that is not bookmarked.
2. Confirm that it's favicon is saved only to the .tabs cache and not to the .bookmarks cache
3. Import bookmarks from HTML file containing some sites that you have not already bookmarked
4. The imported bookmarks will have a default placeholder favicon5. 
5. Tap one of the imported bookmarks to visit that site6. 
6. Return to the bookmarks screen and confirm this bookmark now has a favicon (the same site as visible in tabs view)


<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

amddg44 committed:
Fix for when visiting sites ensuring that only favicons for bookmarked sites are saved to the bookmarks cache

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -706,6 +706,32 @@ extension BookmarksCoreDataStorage {
+            let fetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+            fetchRequest.fetchLimit = 1
+
+            fetchRequest.predicate = NSPredicate(format: "%K CONTAINS %@",
```
Won't this match "tech.com" when searching for "ech.com" ?

bwaresiak commented on Core/BookmarksCoreDataStorage.swift:
```diff
@@ -706,6 +706,32 @@ extension BookmarksCoreDataStorage {
+            fetchRequest.fetchLimit = 1
+
+            fetchRequest.predicate = NSPredicate(format: "%K CONTAINS %@",
+                    #keyPath(BookmarkManagedObject.url.host),
```
Also, I don't think this key path will work with SQLite store, as that store does not treat URLs as "objects", but uses `NSCoreDataStringSearch` function to match strings in DB columns. This probably will work in tests, where we are using in memory store - which is confusing, so we should better move to use SQLite-backed store in tests as well.

Nonetheless, what I mean is that we don't really narrow this to host only, so trying to match "test.com" in "https://example.com/index.html?q=atest.com" will most likely return true.

amddg44 replied:
Good point @bwaresiak. I took a look for an alternative and I see @THISISDINOSAUR  already wrote the logic for checking domains [here](https://github.com/duckduckgo/iOS/blob/65995ec09fab1a14321375431d0fc225c3161a99/DuckDuckGo/BookmarksManager.swift#L185)

I've taken the heart of this logic and added to / encapsulated within `BookmarksCoreDataStorage` and replaced the CoreData call

@THISISDINOSAUR I'm happy to refactor this in a separate PR?  I didn't want to include in this one + wanted to keep this PR on point.

Just one notable for all; I wrapped this new function with `MainActor` - when `Favicons` was calling this method I was consistently getting crashes while testing.  Just mentioning as perhaps its related to the crashes that were mentioned during NativeApps call

THISISDINOSAUR replied:
What kind of refactoring, is there anything besides the tests?

amddg44 committed:
Changes to containsDomain logic

bwaresiak commented on DuckDuckGoTests/FaviconsTests.swift:
```diff
@@ -98,11 +98,34 @@ class FaviconsTests: XCTestCase {
+        let fileName = "MockFiles/DDG-icon_256x256.png"
+
+        guard let imageData = try? FileLoader().load(fileName: fileName, fromBundle: Bundle(for: type(of: self))),
+              let image = UIImage(data: imageData),
```
Can we use something that is already embedded in the app e.g. `UIImage(named: "Logo")` ?

amddg44 replied:
Ah, I didn't know about Logo - thanks for letting me know.  I'm happy to do this if you prefer; however if the `Logo` filename is every changed then the test will break.  The logo I added to mockdata is for the test only.  If we have the unit test use production `Logo`, it would be similar to creating a UnitTest that tests production data (although I can appreciate that in this case the `Logo` name is unlikely to change as often compared to something like prod JSON data).  So it's a small code smell but not too smelly so I don't have too strong an opinion

bwaresiak replied:
Logo image is already used in `loadFaviconSync` so it is related to the problem domain we are touching here - I think any image form that domain would do, logo is just the first one I've spotted. :)

amddg44 replied:
No problem, I've made that change :)

bwaresiak commented on DuckDuckGo/Favicons.swift:
```diff
@@ -35,7 +35,8 @@ public class Favicons {
         static let appUrls = AppUrls()
         static let targetImageSizePoints: CGFloat = 64
-        
+        static let bookmarksCoreDataStorage = BookmarksCoreDataStorage.shared
```
Can we treat this as a dependency passed through the init?

amddg44 replied:
I deleted my last comment and I've made this change. It would be nice in a future issue to remove the singleton so this could be DI'd as I can't use the MockBookmarksCoreDataStore for unit tests

bwaresiak replied:
Why - do you have to use `Favicons.shared` in the tests? Why not create new instance of that just for the tests?

Edit: to be more clear :) Singletons we create have no limitation on init - so in practice it is possible to instantiate these multiple times (which obviously should not happen when executing the app), but this gives us flexibility in tests. In case of unit tests, we can instantiate such class and test functionality in isolation. Obviously this won't work (or would be more difficult) for Integration or UI testing.

amddg44 replied:
I updated my test and all the pre-existing tests to no longer use the singleton

bwaresiak commented on DuckDuckGo/Favicons.swift:
```diff
@@ -138,14 +139,15 @@ public class Favicons {
+              let options = kfOptions(forDomain: domain, usingCache: .bookmarks) else { return }
+
+        if !isFaviconCachedForBookmarks(forDomain: domain, resource: resource) {
+            loadFaviconForBookmarks(forDomain: domain)
```
I don't think this approach is optimal, as we query all bookmarks on each navigation. I'd try to look for a ways to optimize that, like having some caching layer for bookmarks or leveraging existence of tabs cache (so call above code only in case if we insert into tabs cache - as that won't repeat the calls for subsequent navigations on that domain). There are definitely other ideas to consider as well.

amddg44 replied:
I thought about this and tested and I didn't notice any impact to the user experience.  Also I found `BookmarksSearchStore` which unfortunately is not in Core but when looking at this I saw this constantly makes the same call to `BookmarksCoreDataStorage` when using the app so I presumed that this trade off was acceptable.  Every time the address bar is tapped in the app it makes this CoreData call for all bookmarks.

Assuming this trade off is not acceptable is there an existing data cache structure within the app already (data cache) I could potentially leverage for this or (not)?  I can create an enhancement to the original bookmarks import export so that the domain was separated from the bookmark URL on import and cached to disk in some manner + hopefully leverage something the app already has.

Also if we go this route is this worth discussing if MacOS has this and if that can be shared in BSK?

bwaresiak replied:
I think the use case for Search and for this is completely different:
 - In case of search: We cache bookmarks and the result of the query is (nearly) always used when user is typing something in autocomplete to present results. There isn't much happening other than that - so it is not affecting user flow but is part of it.
 - In case of favicon we want to apply it when needed - so when we navigate to a domain and obtain info about favicon related to that domain, then we query bookmarks to see if we need to add this favicon to the cache. User does not need that for his current flow and there is plenty of other things we are doing when loading pages, so I'd rather optimize for using less resources and making less actions. 

Much better would be to do that once per domain (host). Best solution would be to Cache bookmarks (actually - just hosts) in memory and invalidate that cache when there are changes to the store (e.g. on bookmarks did save notification).

I wouldn't think about macOS here too deeply as:
- Desktop has much more resources than mobile.
- Memory separation model between apps on the desktop is different than on mobile, so we may not want to keep bookmark URLs in memory all the time there.

amddg44 replied:
I've taken a look / spent some time on this and what do you think about moving BookmarksCachingSearch to Core and utilising it's cache for favicons as well?  It seems like this could be shared but you know the app better so happy to defer to you 🙂

bwaresiak replied:
I think it makes sense. To be on the same page (as I'm out next week and someone will need to pick this up from here):
 - We should be able to instantiate `BookmarksCachingSearch` lazily when needed, then discard it once we receive `dataDidChange` notification from `BookmarksCoreDataStorage`. It should be fine to just keep that in memory as we already do every time user taps URL bar.
 - It probably is also a good idea to change Autosuggest to use the same instance.

amddg44 committed:
Modified to pass in bookmarksCoreDataStorage through init

amddg44 committed:
Updated to use existing embedded image + no longer use singleton in the test

amddg44 committed:
BookmarksCachingSearch.swift moved to Core

amddg44 committed:
- CoreDependencyProvider.swift created to provide BookmarksCachingSearch
- BookmarksCachingSearch.swift updated to provide in memory cache
- Favicons.swift and ViewControllers updated to use BookmarksCachingSearch.swift from CoreDependencyProvider.swift

amddg44 committed:
Removed circular dependency between Favicons.swift and NotFoundCachingDownloader.swift

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/fix-bookmarks-favicon-caching

THISISDINOSAUR commented on DuckDuckGo/BookmarksViewController.swift:
```diff
@@ -357,6 +357,10 @@ class BookmarksViewController: UITableViewController {
     func importBookmarks(fromHtml html: String) {
+        // pre-emptively deregisterForNotifications so that bookmarksCachingSearch is not saturated with notification events
+        // and constantly rebuilding while bookmarks are being imported (bookmark files could be very large)
+        bookmarksCachingSearch.deregisterForNotifications()
```
Very very minor, but I'd prefer if this was moved from the view controller since it feels like something the view controller shouldn't have to know about. E.g. managed by BookmarksImporter or even just behind a method on bookmarksCachingSearch (like `prepareForImport()` or something).

THISISDINOSAUR replied:
Also wondering does the import trigger loads of notifications in any other bookmarksCachingSearch instances that are around? I mean it won't right now since I think the same shared object is used everywhere? But in future it would be easy for someone to not release this and trip up 

How about importing firing notifications before starting and on completion that bookmarksCachingSearch can listen to and then manage what it does with that information itself?

amddg44 replied:
> Very very minor, but I'd prefer if this was moved from the view controller since it feels like something the view controller shouldn't have to know about. E.g. managed by BookmarksImporter or even just behind a method on bookmarksCachingSearch (like `prepareForImport()` or something).

Sounds good + thanks for the suggestion!

amddg44 replied:
> Also wondering does the import trigger loads of notifications in any other bookmarksCachingSearch instances that are around? I mean it won't right now since I think the same shared object is used everywhere? But in future it would be easy for someone to not release this and trip up
> 
> How about importing firing notifications before starting and on completion that bookmarksCachingSearch can listen to and then manage what it does with that information itself?

Sounds great + I've done it this way :slightly_smiling_face:

THISISDINOSAUR commented on Core/FaviconsHelper.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+import Kingfisher
+
+struct FaviconsHelper {
```
Heads up @Bunn in case I forget. I can't remember what you called your Favicon thing, but we might want to unify it with this when merging

THISISDINOSAUR commented on Core/CoreDependencyProvider.swift:
```diff
@@ -0,0 +1,32 @@
+
+/// Provides dependencies for objects that are not directly instantiated
+/// through `init` call
+public class CoreDependencyProvider: CoreDependencyProvided {
```
This is interesting, how do you plan on using it in future?

amddg44 replied:
I saw AppDependencyProvider already in place and thought Core needed something similar in this case; ideally we would move our singletons such as Favicons to this DependencyProvider for DI in the future

amddg44 committed:
Decouple ViewController and CachingSearch import events

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/fix-bookmarks-favicon-caching

# Conflicts:
#	DuckDuckGo/Favicons.swift

THISISDINOSAUR reviewed (approved):
LGTM!
